### PR TITLE
feat: Implement CALLDATACOPY opcode

### LINKED ISSUE #37: Opcode: CALLDATACOPY

### PR DESCRIPTION
**Pull request by Danikim01**
closes #37 
This PR implements the CALLDATACOPY opcode

### CODE REVIEW

File: src/codegen/operations.rs [hunk 1]
```diff
@@ -71,9 +71,225 @@ pub fn generate_code_for_op<'c>(
...
+        .result(0)?
+        .into();
+
+    overflow_block.append_operation(
+        ods::llvm::intr_memset(
+            context,
+            zero_fill_dest,
+            zero_value,
+            zero_fill_size,
+            IntegerAttribute::new(IntegerType::new(context, 1).into(), 1),
```
Comment (MegaRedHand) [inline src/codegen/operations.rs:282]: This marks the value as volatile? Why?

File: src/main.rs [hunk 2]
```diff
@@ -1,7 +1,8 @@
 use std::path::PathBuf;
 
 use evm_mlir::{context::Context, executor::Executor, program::Program, syscall::SyscallContext};
-
+// use num_bigint::BigUint;
+// use evm_mlir::program::Operation;
```
Comment (MegaRedHand) [inline src/main.rs:5]: There's commented code here

File: src/main.rs [hunk 3]
```diff
@@ -29,4 +30,37 @@ fn main() {
     let result = executor.execute(&mut context, initial_gas);
 
     println!("Execution result: {result}");
+
```
Comment (MegaRedHand) [inline src/main.rs:33]: And here

File: src/syscall.rs [hunk 4]
```diff
@@ -89,14 +89,14 @@ impl ExecutionResult {
 pub struct SyscallContext {
     /// The memory segment of the EVM.
     /// For extending it, see [`Self::extend_memory`]
-    memory: Vec<u8>,
+    pub memory: Vec<u8>,
```
Comment (MegaRedHand) [inline src/syscall.rs:92]: None of these should be public

File: tests/operations.rs [hunk 5]
```diff
@@ -98,6 +98,168 @@ pub fn biguint_256_from_bigint(value: BigInt) -> BigUint {
     }
 }
 
+#[test]
+fn test_calldatacopy() {
```
Comment (MegaRedHand) [inline tests/operations.rs:102]: Please move tests to `tests/evm.rs`, and use the `Evm` API.

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

File: src/syscall.rs [hunk 6]
```diff
@@ -317,6 +350,45 @@ pub(crate) mod mlir {
...
+                FlatSymbolRefAttribute::new(mlir_ctx, symbols::GET_CALLDATA_PTR),
+                &[syscall_ctx],
+                &[ptr_type],
+                location,
+            ))
+            .result(0)?;
+        Ok(value.into())
+    }
+
+    pub(crate) fn get_memory_pointer_syscall<'c>(
```
Comment (MegaRedHand) [inline src/syscall.rs:373]: `extend_memory` returns the memory pointer. That's why this one wasn't added.

File: src/syscall.rs [hunk 7]
```diff
@@ -317,6 +350,45 @@ pub(crate) mod mlir {
         Ok(value.into())
     }
 
+    /// Returns a pointer to the start of the calldata
+    pub(crate) fn get_calldata_ptr_syscall<'c>(
```
Comment (MegaRedHand) [inline src/syscall.rs:354]: This one can be removed, since it's duplicate

File: src/codegen/operations.rs [hunk 8]
```diff
@@ -78,11 +78,247 @@ pub fn generate_code_for_op<'c>(
...
+        .unwrap()
+        .into();
+
+    let size = ok_block
+        .append_operation(arith::trunci(size, uint32.into(), location))
+        .result(0)
+        .unwrap()
+        .into();
+
+    //required size = call_data_offset + size
```
Comment (MegaRedHand) [inline src/codegen/operations.rs:146]:
```suggestion
    // required size = dest_offset + size
```

File: src/codegen/operations.rs [hunk 9]
```diff
@@ -78,11 +78,247 @@ pub fn generate_code_for_op<'c>(
...
+        .into();
+
+    // Check that required_calldata_size <= calldatasize
+
+    let calldata_size = op_ctx.get_calldata_size_syscall(&ok_block, location)?;
+
+    let remaining_calldata_size = ok_block
+        .append_operation(arith::subi(calldata_size, call_data_offset, location))
+        .result(0)?
+        .into();
```
Comment (MegaRedHand) [inline src/codegen/operations.rs:174]: This can be negative, right? Are we handling that case?

**Comment by Danikim01** [comment]
> closes #37 This PR implements the CALLDATACOPY opcode

with the new changes closes #242

File: src/codegen/operations.rs [hunk 10]
```diff
@@ -81,12 +81,189 @@ pub fn generate_code_for_op<'c>(
...
+    let continue_memory_block = region.append_block(Block::new(&[]));
+    extend_memory(
+        op_ctx,
+        &ok_block,
+        &continue_memory_block,
+        region,
+        required_memory_size,
+        gas_cost::CALLDATACOPY,
+    )?;
+    let memory_ptr = return_memory_pointer(op_ctx, &continue_memory_block, location)?;
```
[the remark this answers is not in the document]
Comment (Danikim01) [reply src/codegen/operations.rs]: resolved

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**